metrics_api/test/opentelemetry/metrics/meter_test.rb

```diff
@@ -25,49 +25,259 @@
+        _(-> { meter.create_histogram('an_instrument', unit: invalid_unit) })
+          .must_raise(instrument_unit_error, "should have raised for unit=#{invalid_unit.inspect}")
+      end
+    end
+
+    it 'raises an error when description is invalid' do
+      invalid_descriptions.each do |invalid_description|
+        _(-> { meter.create_histogram('an_instrument', description: invalid_description) })
+          .must_raise(instrument_description_error, "should have raised for description=#{invalid_description.inspect}")
+      end
+    end
+
+    it 'creates and returns an instance of Histogram instrument' do
+      assert(meter.create_histogram('an_instrument')
+        .is_a?(OpenTelemetry::Metrics::Instrument::Histogram))
+    end
+
+    it 'logs a warning message when instrument with same name already exists' do
+      OpenTelemetry::TestHelpers.with_test_logger do |log_stream|
+        meter.create_histogram('an_instrument')
+        meter.create_histogram('AN_INSTRUMENT')
+        _(log_stream.string).must_match(/duplicate instrument registration occurred for instrument an_instrument/)
+      end
+    end
+  end
+
+  describe '#create_up_down_counter' do
+    it 'raises an error when name does not match pattern' do
+      invalid_names.each do |invalid_name|
+        _(-> { meter.create_up_down_counter(invalid_name) })
+          .must_raise(instrument_name_error, "should have raised for name=#{invalid_name.inspect}")
+      end
+    end
+
+    it 'raises an error when unit is invalid' do
+      invalid_units.each do |invalid_unit|
+        _(-> { meter.create_up_down_counter('an_instrument', unit: invalid_unit) })
+          .must_raise(instrument_unit_error, "should have raised for unit=#{invalid_unit.inspect}")
+      end
+    end
+
+    it 'raises an error when description is invalid' do
+      invalid_descriptions.each do |invalid_description|
+        _(-> { meter.create_up_down_counter('an_instrument', description: invalid_description) })
+          .must_raise(instrument_description_error, "should have raised for description=#{invalid_description.inspect}")
+      end
+    end
+
+    it 'creates and returns an instance of UpDownCounter instrument' do
+      assert(meter.create_up_down_counter('an_instrument')
+        .is_a?(OpenTelemetry::Metrics::Instrument::UpDownCounter))
+    end
+
+    it 'logs a warning message when instrument with same name already exists' do
+      OpenTelemetry::TestHelpers.with_test_logger do |log_stream|
+        meter.create_up_down_counter('an_instrument')
+        meter.create_up_down_counter('AN_INSTRUMENT')
+        _(log_stream.string).must_match(/duplicate instrument registration occurred for instrument an_instrument/)
+      end
+    end
+  end
+
+  describe '#create_observable_counter' do
+    it 'raises an error when name does not match pattern' do
+      invalid_names.each do |invalid_name|
+        _(-> { meter.create_observable_counter(invalid_name, callback: -> {}) })
```

Comment:
**Not related to this PR, but I can include:**

I noticed the keyword arg `callback:` is required. So here I'm passing it even though I don't need it for the test case.

[From the specification](https://opentelemetry.io/docs/reference/specification/metrics/api/#asynchronous-instrument-api), it says:

> Users can provide `callback` functions, but it is up to their discretion. Therefore, this API MUST be structured to accept a variable number of `callback` functions, including none.

It **doesn't say** _"MUST NOT obligate a user to provide one"_ like for the other parameters, but it says _"including none"_.

I wonder if _"including none"_ would translate to optional in Ruby with a default empty `[]` array. What do you think?

In doc terms:

```rb
# @param callback [optional Proc, Array<Proc>]
```